### ISSUE TITLE
PersonFacade.count/getIndexList: Slow query

### ISSUE THREAD

**Issue by StefanKock** [hzi-braunschweig/SORMAS-Project#6170]
### Bug Description

Found by @AndyBakcsy-she when checking after #4922: Loading persons can again behave too slow. First guess: This feature is extended to also include persons that relate to travel entries and immunizations. Maybe this causes the problem.

### Steps to Reproduce

1. Open view persons (see system details below)

### Expected Behavior

Return results in reasonable time (getIndexList took ~14min in that example)

### System Details

* SORMAS version: 1.62.0-SNAPSHOT
* Server URL: https://test.sormas.netzlink.com
* User Role: admin surveillance supervisor, User AB_ContSup

### Additional Information

Slow queries:

<details>
<summary>Person.count</summary>

```sql
SELECT count(DISTINCT person0_.id) AS col_0_0_
FROM Person person0_
LEFT OUTER JOIN cases cases1_ ON person0_.id=cases1_.person_id
LEFT OUTER JOIN Contact contacts2_ ON cases1_.id=contacts2_.caze_id
LEFT OUTER JOIN Contact contacts3_ ON person0_.id=contacts3_.person_id
LEFT OUTER JOIN cases case4_ ON contacts3_.caze_id=case4_.id
LEFT OUTER JOIN Contact contacts5_ ON case4_.id=contacts5_.caze_id
LEFT OUTER JOIN EventParticipant eventparti6_ ON person0_.id=eventparti6_.person_id
LEFT OUTER JOIN EVENTS event7_ ON eventparti6_.event_id=event7_.id
LEFT OUTER JOIN LOCATION location8_ ON event7_.eventLocation_id=location8_.id
LEFT OUTER JOIN users user9_ ON event7_.reportingUser_id=user9_.id
LEFT OUTER JOIN users user10_ ON event7_.responsibleUser_id=user10_.id
LEFT OUTER JOIN cases case11_ ON eventparti6_.resultingCase_id=case11_.id
LEFT OUTER JOIN Contact contacts12_ ON case11_.id=contacts12_.caze_id
LEFT OUTER JOIN immunization immunizati13_ ON person0_.id=immunizati13_.person_id
LEFT OUTER JOIN users user14_ ON immunizati13_.reportingUser_id=user14_.id
LEFT OUTER JOIN Region region15_ ON immunizati13_.responsibleRegion_id=region15_.id
LEFT OUTER JOIN travelentry travelentr16_ ON person0_.id=travelentr16_.person_id
LEFT OUTER JOIN cases case17_ ON travelentr16_.resultingCase_id=case17_.id
LEFT OUTER JOIN users user18_ ON case17_.reportingUser_id=user18_.id
LEFT OUTER JOIN Region region19_ ON case17_.responsibleRegion_id=region19_.id
LEFT OUTER JOIN Region region20_ ON case17_.region_id=region20_.id
LEFT OUTER JOIN users user21_ ON travelentr16_.reportingUser_id=user21_.id
LEFT OUTER JOIN Region region22_ ON travelentr16_.responsibleRegion_id=region22_.id
LEFT OUTER JOIN Region region23_ ON travelentr16_.pointOfEntryRegion_id=region23_.id
LEFT OUTER JOIN LOCATION location24_ ON person0_.address_id=location24_.id
LEFT OUTER JOIN Region region25_ ON location24_.region_id=region25_.id
LEFT OUTER JOIN District district26_ ON location24_.district_id=district26_.id
LEFT OUTER JOIN Community community27_ ON location24_.community_id=community27_.id
WHERE (cases1_.region_id=156
       OR cases1_.responsibleRegion_id=156
       OR contacts2_.reportingUser_id=$1
       OR contacts2_.contactOfficer_id=$2
       OR contacts2_.region_id=$3
       OR cases1_.reportingUser_id=2177110
       OR cases1_.surveillanceOfficer_id=2177110
       OR cases1_.caseOfficer_id=2177110)
  AND cases1_.deleted=FALSE
  OR (contacts3_.reportingUser_id=$4
      OR contacts3_.contactOfficer_id=$5
      OR contacts3_.region_id=$6
      OR case4_.region_id=156
      OR case4_.responsibleRegion_id=156
      OR contacts5_.reportingUser_id=$7
      OR contacts5_.contactOfficer_id=$8
      OR contacts5_.region_id=$9
      OR case4_.reportingUser_id=2177110
      OR case4_.surveillanceOfficer_id=2177110
      OR case4_.caseOfficer_id=2177110)
  AND contacts3_.deleted=FALSE
  OR (location8_.region_id=$10
      OR case11_.region_id=156
      OR case11_.responsibleRegion_id=156
      OR contacts12_.reportingUser_id=$11
      OR contacts12_.contactOfficer_id=$12
      OR contacts12_.region_id=$13
      OR case11_.reportingUser_id=2177110
      OR case11_.surveillanceOfficer_id=2177110
      OR case11_.caseOfficer_id=2177110
      OR eventparti6_.region_id=156
      OR location8_.region_id=$14
      OR user9_.id=$15
      OR user10_.id=$16)
  AND eventparti6_.deleted=FALSE
  OR ((user14_.id IS NOT NULL)
      AND user14_.uuid=$17
      OR region15_.id=156)
  AND immunizati13_.deleted=FALSE
  OR ((user21_.id IS NOT NULL)
      AND user21_.uuid=$18
      OR region22_.id=156
      OR region23_.id=156
      OR (user18_.id IS NOT NULL)
      AND user18_.uuid=$19
      OR region19_.id=156
      OR region20_.id=156)
  AND travelentr16_.deleted=FALSE

-- 2021-07-20 11:40:16.336 GMT [48] LOG:  duration: 421406.024 ms  execute
-- 2021-07-20 11:40:16.336 GMT [48] DETAIL:  parameters: $1 = '2177110', $2 = '2177110', $3 = '156', $4 = '2177110', $5 = '2177110', $6 = '156', $7 = '2177110', $8 = '2177110', $9 = '156', $10 = '156', $11 = '2177110', $12 = '2177110', $13 = '156', $14 = '156', $15 = '2177110', $16 = '2177110', $17 = 'VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU', $18 = 'VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU', $19 = 'VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU'
```

</details>

<details>
<summary>Person.getIndexList</summary>

```sql
SELECT DISTINCT person0_.uuid AS col_0_0_,
                person0_.firstName AS col_1_0_,
                person0_.lastName AS col_2_0_,
                person0_.approximateAge AS col_3_0_,
                person0_.approximateAgeType AS col_4_0_,
                person0_.birthdate_dd AS col_5_0_,
                person0_.birthdate_mm AS col_6_0_,
                person0_.birthdate_yyyy AS col_7_0_,
                person0_.sex AS col_8_0_,
                district2_.name AS col_9_0_,
                location1_.street AS col_10_0_,
                location1_.houseNumber AS col_11_0_,
                location1_.postalCode AS col_12_0_,
                location1_.city AS col_13_0_,

  (SELECT personcont43_.contactInformation
   FROM PersonContactDetail personcont43_
   WHERE personcont43_.person_id=person0_.id
     AND personcont43_.primaryContact=TRUE
     AND personcont43_.personContactDetailType=$1) AS col_14_0_,

  (SELECT personcont44_.contactInformation
   FROM PersonContactDetail personcont44_
   WHERE personcont44_.person_id=person0_.id
     AND personcont44_.primaryContact=TRUE
     AND personcont44_.personContactDetailType=$2) AS col_15_0_,
                person0_.changeDate AS col_16_0_,
                CASE
                    WHEN 0=1
                         OR (user4_.id IS NOT NULL)
                         AND user4_.uuid='VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU'
                         OR region5_.id=156
                         OR region6_.id=156
                         OR (user12_.id IS NOT NULL)
                         AND user12_.uuid='VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU'
                         OR region13_.id=156
                         OR region10_.id=156
                         OR region11_.id=156
                         OR (user17_.id IS NOT NULL)
                         AND user17_.uuid='VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU'
                         OR region18_.id=156
                         OR region21_.id=156 THEN TRUE
                    ELSE FALSE
                END AS col_17_0_
FROM Person person0_
LEFT OUTER JOIN LOCATION location1_ ON person0_.address_id=location1_.id
LEFT OUTER JOIN District district2_ ON location1_.district_id=district2_.id
LEFT OUTER JOIN cases cases3_ ON person0_.id=cases3_.person_id
LEFT OUTER JOIN users user4_ ON cases3_.reportingUser_id=user4_.id
LEFT OUTER JOIN Region region5_ ON cases3_.responsibleRegion_id=region5_.id
LEFT OUTER JOIN Region region6_ ON cases3_.region_id=region6_.id
LEFT OUTER JOIN Contact contacts7_ ON cases3_.id=contacts7_.caze_id
LEFT OUTER JOIN Contact contacts8_ ON person0_.id=contacts8_.person_id
LEFT OUTER JOIN cases case9_ ON contacts8_.caze_id=case9_.id
LEFT OUTER JOIN Region region10_ ON case9_.responsibleRegion_id=region10_.id
LEFT OUTER JOIN Region region11_ ON case9_.region_id=region11_.id
LEFT OUTER JOIN users user12_ ON contacts8_.reportingUser_id=user12_.id
LEFT OUTER JOIN Region region13_ ON contacts8_.region_id=region13_.id
LEFT OUTER JOIN cases case14_ ON contacts8_.caze_id=case14_.id
LEFT OUTER JOIN Contact contacts15_ ON case14_.id=contacts15_.caze_id
LEFT OUTER JOIN EventParticipant eventparti16_ ON person0_.id=eventparti16_.person_id
LEFT OUTER JOIN users user17_ ON eventparti16_.reportingUser_id=user17_.id
LEFT OUTER JOIN Region region18_ ON eventparti16_.region_id=region18_.id
LEFT OUTER JOIN EVENTS event19_ ON eventparti16_.event_id=event19_.id
LEFT OUTER JOIN LOCATION location20_ ON event19_.eventLocation_id=location20_.id
LEFT OUTER JOIN Region region21_ ON location20_.region_id=region21_.id
LEFT OUTER JOIN EVENTS event22_ ON eventparti16_.event_id=event22_.id
LEFT OUTER JOIN LOCATION location23_ ON event22_.eventLocation_id=location23_.id
LEFT OUTER JOIN users user24_ ON event22_.reportingUser_id=user24_.id
LEFT OUTER JOIN users user25_ ON event22_.responsibleUser_id=user25_.id
LEFT OUTER JOIN cases case26_ ON eventparti16_.resultingCase_id=case26_.id
LEFT OUTER JOIN Contact contacts27_ ON case26_.id=contacts27_.caze_id
LEFT OUTER JOIN immunization immunizati28_ ON person0_.id=immunizati28_.person_id
LEFT OUTER JOIN users user29_ ON immunizati28_.reportingUser_id=user29_.id
LEFT OUTER JOIN Region region30_ ON immunizati28_.responsibleRegion_id=region30_.id
LEFT OUTER JOIN travelentry travelentr31_ ON person0_.id=travelentr31_.person_id
LEFT OUTER JOIN cases case32_ ON travelentr31_.resultingCase_id=case32_.id
LEFT OUTER JOIN users user33_ ON case32_.reportingUser_id=user33_.id
LEFT OUTER JOIN Region region34_ ON case32_.responsibleRegion_id=region34_.id
LEFT OUTER JOIN Region region35_ ON case32_.region_id=region35_.id
LEFT OUTER JOIN users user36_ ON travelentr31_.reportingUser_id=user36_.id
LEFT OUTER JOIN Region region37_ ON travelentr31_.responsibleRegion_id=region37_.id
LEFT OUTER JOIN Region region38_ ON travelentr31_.pointOfEntryRegion_id=region38_.id
LEFT OUTER JOIN LOCATION location39_ ON person0_.address_id=location39_.id
LEFT OUTER JOIN Region region40_ ON location39_.region_id=region40_.id
LEFT OUTER JOIN District district41_ ON location39_.district_id=district41_.id
LEFT OUTER JOIN Community community42_ ON location39_.community_id=community42_.id
WHERE (cases3_.region_id=156
       OR cases3_.responsibleRegion_id=156
       OR contacts7_.reportingUser_id=$3
       OR contacts7_.contactOfficer_id=$4
       OR contacts7_.region_id=$5
       OR cases3_.reportingUser_id=2177110
       OR cases3_.surveillanceOfficer_id=2177110
       OR cases3_.caseOfficer_id=2177110)
  AND cases3_.deleted=FALSE
  OR (contacts8_.reportingUser_id=$6
      OR contacts8_.contactOfficer_id=$7
      OR contacts8_.region_id=$8
      OR case14_.region_id=156
      OR case14_.responsibleRegion_id=156
      OR contacts15_.reportingUser_id=$9
      OR contacts15_.contactOfficer_id=$10
      OR contacts15_.region_id=$11
      OR case14_.reportingUser_id=2177110
      OR case14_.surveillanceOfficer_id=2177110
      OR case14_.caseOfficer_id=2177110)
  AND contacts8_.deleted=FALSE
  OR (location23_.region_id=$12
      OR case26_.region_id=156
      OR case26_.responsibleRegion_id=156
      OR contacts27_.reportingUser_id=$13
      OR contacts27_.contactOfficer_id=$14
      OR contacts27_.region_id=$15
      OR case26_.reportingUser_id=2177110
      OR case26_.surveillanceOfficer_id=2177110
      OR case26_.caseOfficer_id=2177110
      OR eventparti16_.region_id=156
      OR location23_.region_id=$16
      OR user24_.id=$17
      OR user25_.id=$18)
  AND eventparti16_.deleted=FALSE
  OR ((user29_.id IS NOT NULL)
      AND user29_.uuid=$19
      OR region30_.id=156)
  AND immunizati28_.deleted=FALSE
  OR ((user36_.id IS NOT NULL)
      AND user36_.uuid=$20
      OR region37_.id=156
      OR region38_.id=156
      OR (user33_.id IS NOT NULL)
      AND user33_.uuid=$21
      OR region34_.id=156
      OR region35_.id=156)
  AND travelentr31_.deleted=FALSE
ORDER BY person0_.changeDate DESC
LIMIT $22

-- 2021-07-20 11:54:34.100 GMT [48] LOG:  duration: 857379.978 ms  execute
-- 2021-07-20 11:54:34.100 GMT [48] DETAIL:  parameters: $1 = 'PHONE', $2 = 'EMAIL', $3 = '2177110', $4 = '2177110', $5 = '156', $6 = '2177110', $7 = '2177110', $8 = '156', $9 = '2177110', $10 = '2177110', $11 = '156', $12 = '156', $13 = '2177110', $14 = '2177110', $15 = '156', $16 = '156', $17 = '2177110', $18 = '2177110', $19 = 'VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU', $20 = 'VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU', $21 = 'VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU', $22 = '100'
```

</details>

Analysis: https://explain.dalibo.com/plan/gmS (also contains the query to copy for rework). [node 19](https://explain.dalibo.com/plan/gmS#plan/node/19) brings are remarkable cartesic product.

<details>
<summary>System resources usage while running the analysis</summary>

![grafik](https://user-images.githubusercontent.com/24678567/126336167-422e41b2-300f-448c-98d2-1241d4aacaa2.png)

</details>

**Comment by StefanKock** [comment]
I'm not certain but it seems that several joins on `Contact` produce two times a full cartesic product.

**Comment by StefanKock** [comment]
Findings (probably not directly related to this performance problem, but still worth to look into):

1. Joins for immunizations too complex: Comparison can be done by id on the Join attributes `immunizati28_.reportingUser_id` and `immunizati28_.responsibleRegion_id`. Relates to #4756 
```sql
  OR ((user29_.id IS NOT NULL)
      AND user29_.uuid='VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU'
      OR region30_.id=156)
  AND immunizati28_.deleted=FALSE
```
2. Joins for travel entries too complex: Comparison can be done by id on the Join attributes `travelentr31_.reportingUser_id`, `travelentr31_.responsibleRegion_id`, `travelentr31_.pointOfEntryRegion_id`, `case32_.reportingUser_id`, `case32_.responsibleRegion_id`, `case32_.region_id`. Relates to #6022
```sql
  OR ((user36_.id IS NOT NULL)
      AND user36_.uuid='VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU'
      OR region37_.id=156
      OR region38_.id=156
      OR (user33_.id IS NOT NULL)
      AND user33_.uuid='VQKA2G-M3ZXHZ-T2S6R2-FPFSSASU'
      OR region34_.id=156
      OR region35_.id=156)
```
3. Joins for event participants too complex: Comparison can be done by id on the Join attributes `event22_.reportingUser_id`, `event22_.responsibleUser_id`. New ticket
```sql
  OR (location23_.region_id='156'
      OR case26_.region_id=156
      OR case26_.responsibleRegion_id=156
      OR contacts27_.reportingUser_id='2177110'
      OR contacts27_.contactOfficer_id='2177110'
      OR contacts27_.region_id='156'
      OR case26_.reportingUser_id=2177110
      OR case26_.surveillanceOfficer_id=2177110
      OR case26_.caseOfficer_id=2177110
      OR eventparti16_.region_id=156
      OR location23_.region_id= '156'
      OR user24_.id= '2177110'
      OR user25_.id= '2177110')
```

**Comment by StefanKock** [comment]
I removed the travel entries and immunizations parts and the problem persists (it actually get linear worse with more joins, but it's not the cause).
Analysis: https://explain.dalibo.com/plan/CZX

**Comment by StefanKock** [comment]
I thought about reducing the JOIN complexity here by fetching the person ids per entity of interest and then UNION the result.

<details>
<summary>Pseudo code for possible new query structure</summary>

```sql
SELECT DISTINCT attributes FROM person

-- For plain attributes
LEFT OUTER JOIN location
LEFT OUTER JOIN district

-- For calculated attribute isInJurisdiction
LEFT OUTER JOIN case
LEFT OUTER JOIN contact
LEFT OUTER JOIN eventparticipant
LEFT OUTER JOIN location
LEFT OUTER JOIN event
LEFT OUTER JOIN immunization
LEFT OUTER JOIN travelentry

WHERE person.id IN
(
SELECT person.id FROM person -- JOIN WHERE case userFilter
UNION
SELECT person.id FROM person -- JOIN WHERE contact userFilter
UNION
SELECT person.id FROM person -- JOIN WHERE eventparticipant userFilter
UNION
SELECT person.id FROM person -- JOIN WHERE immunization userFilter
UNION
SELECT person.id FROM person -- JOIN WHERE travelentry userFilter
)
```

</details>

**Comment by StefanKock** [comment]
I think I found the reason for the data explosion: some cases have hundreds of contacts.
```sql
SELECT caze_id, count(*) FROM contact GROUP BY caze_id ORDER BY COUNT DESC;

 caze_id | count
---------+-------
     314 |   544
 2270665 |   224
         |    97
     454 |    27
    1588 |    13
 2282909 |     8
     145 |     5
 2283801 |     5
 1830515 |     4
```

**Comment by StefanKock** [comment]
Concluded with @sergiupacurariu and @BarnaBartha that we found some problems that can be fixed within this ticket:
1. Reduce joins for jurisdiction checks on old code (`Case, Contact, Eventparticipant`) and new code (`TravelEntry, Immunization`).
2. Reduce applied filters if query is filtered to one partition
3. Add immunization + travel entry association filters to person directory
4. Show warning for restricted users when searching for all associations that it might be a longer query

Points 1 to 4 of https://github.com/hzi-braunschweig/SORMAS-Project/issues/6170#issuecomment-883472945 are addressed to go into 1.62.0 to help to mitigate the problem, but the cause (https://github.com/hzi-braunschweig/SORMAS-Project/issues/6170#issuecomment-884669944) will have to be addressed by #6188.

**Comment by StefanKock** [comment]
I'll gonna run the analysis again once this is deployed on test.sormas

**Comment by StefanKock** [comment]
Retest on reported system: The behaviour is expected, it still has signifant runtime of 4min but was reduced thx to less joins. https://explain.dalibo.com/plan/axk

**Comment by AndyBakcsy-she** [comment]
The issue does not reproduce anymore on the environment where this was reported ( test.sormas.netzlink.com) 
Loading of the Persons page took less than 5 seconds.
